### PR TITLE
HAAR-1343 update external user email

### PR DESCRIPTION
**Pull request by AlTurner-MOJ**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserSyncController.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ */
+@RestController
+class AuthUserSyncController(
+  val authUserSyncService: AuthUserSyncService,
+  val telemetryClient: TelemetryClient,
+) {
+
+  @PutMapping("/api/authuser/sync/{username}/email")
+  @PreAuthorize("hasAnyRole('ROLE_MAINTAIN_OAUTH_USERS', 'ROLE_AUTH_GROUP_MANAGER')")
+  fun emailUpdate(@PathVariable username: String, @RequestBody emailUpdate: UserEmailUpdate) {
```
Comment (ravmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserSyncController.kt:28]: Add swagger docs

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/maintain/AuthUserSyncService.kt [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+@Service
+@Transactional
+class AuthUserSyncService(
+  private val userRepository: UserRepository,
+) {
+  fun syncUserEmail(username: String, emailUpdate: UserEmailUpdate) {
+    val user = userRepository.findByUsername(username).orElseThrow { UsernameNotFoundException("User with name $username not found") }
+    user.email = emailUpdate.email
+    user.username = emailUpdate.username
+    user.verified = false
```
Comment (ravmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/maintain/AuthUserSyncService.kt:18]: why is it false?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserSyncController.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ * Although part of the Auth API, these end points are intended to be used exclusively from hmpps-manage-users-api
+ * to keep external user data in sync, and therefore are not part of the published API.
+ */
+@RestController
+class AuthUserSyncController(
+  val authUserSyncService: AuthUserSyncService,
+  val telemetryClient: TelemetryClient,
+) {
+
+  @PutMapping("/api/authuser/sync/{username}/email")
```
Comment (ravmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserSyncController.kt:26]: do we need to add role for endpoints to access internal services?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserSyncController.kt (hunk 1 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserSyncController.kt:28]: This was discussed with @karenmillermoj - the intention of this (and other) sync end points is that they will only be used from manage-users-api for the purpose of keeping user related data sychronised between the Auth and external-users databases. These end points will not form part of the 'published' Auth API, which is why there is no Swagger documentation.
The roles on this end point are the same as those required to use the single client context (amend a user email)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/maintain/AuthUserSyncService.kt (hunk 2 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/maintain/AuthUserSyncService.kt:18]: This is a little bit of business logic which you are right to question. I wanted to make this end point essentially dumb in the sense that it just does what the client asks (and assumes that the client has completed any necessary validations/checks). I debated whether to allow the client to pass the verified flag in, but the only use case for this is altering a user's email, and we always set the verified flag to false when this happens.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserSyncController.kt (hunk 3 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/api/AuthUserSyncController.kt:26]: Possibly - I left the role the same as that required to access the single intended client use case (amend user email). Did not feel like adding a specific separate role for this should be necessary?

**Review by ravmoj — APPROVED** [review]
lgtm

**Pull request merged**